### PR TITLE
[inductor] Use python constants in IndexPropagation

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #115435
* #121924
* __->__ #122031

In the next PR I have the IR `ops.neg(ops.constant(0.0, torch.float32))`
which should be folded to `ops.constant(-0.0, torch.float32)` but it seems that
`sympy.Float(-0.0)` doesn't respect the sign of the zero and so we instead
get a positive zero constant.

Here, I work around this by doing the constant folding with python arithmetic
which does respect signed zeros.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Review by lezcano — APPROVED** [review]

File: torch/_inductor/index_propagation.py [hunk 1]
```diff
@@ -125,11 +127,14 @@ def remainder(x: TypedExpr, y: TypedExpr) -> Optional[TypedExpr]:
         result_type = torch.promote_types(x.dtype, y.dtype)
         if not is_integer_dtype(result_type):
             return NotImplemented
+
+        x_expr = sympy.sympify(x.expr)
+        y_expr = sympy.sympify(y.expr)
```
[the remark this answers is not in the document]
Comment (peterbell10) [reply torch/_inductor/index_propagation.py:132]: If this is the right function, then it doesn't allow expressions with free symbols so kind of pointless.
https://github.com/pytorch/pytorch/blob/f2f8eeea944f5cc6dd6f907a0c78067f73e0ad9c/torch/utils/_sympy/value_ranges.py#L28-L29

It also seems to be much slower than `sympy.sympify`:
```
import sympy
s = sympy.Float(0.0)
%timeit sympy.sympify(s)
# 232 ns ± 0.64 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
%timeit simple_sympify(s)
# 1.53 µs ± 1.64 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```

[the remark this answers is not in the document]
Comment (lezcano) [reply torch/_inductor/index_propagation.py:132]: It looks like `sympy.sympify` it has an early return for sympy objects. For python objects the custom implementation it's either comparable for float constants or twice as fast for integers.
We might want to add an early return tho.

[the remark this answers is not in the document]
Comment (lezcano) [reply torch/_inductor/index_propagation.py:132]: but yeah, if it's not just for constants, then that function cannot be used.

**Comment by Skylion007** [comment]
Can we also file an issue in Sympy to get this fixed for a future release? Seems like a bug on their end.

**Comment by peterbell10** [comment]
Perhaps @asmeurer has more context, but it seems to be intended behavior. From the [sympy docs](https://docs.sympy.org/latest/modules/core.html#sympy.core.numbers.Float):
> Zero in Float only has a single value. Values are not separate for positive and negative zeroes.

**Comment by asmeurer** [comment]
[mpmath doesn't distinguish positive and negative zero](https://mpmath.org/doc/current/technical.html#:~:text=(Mpmath%20does%20not%20currently%20distinguish%20between%20positive%20and%20negative%20zero.)), and so sympy doesn't either since Float is just a wrapper around mpf. I don't know the original motivation for that decision. I imagine in SymPy Float(-0.) could cause a lot of headaches and so the community would probably be hesitant to add it. The problems -0 generally solves are typically solved in other ways in SymPy.

**Pull request closed without merging**